File: Problems/5. Longest Palindromic Substring/longest-palindromic-substring.ts
//Solution 1: 69ms 45.23 MB
function expandAroundCenter(s: string, left: number, right: number) {
  while (left >= 0 && right < s.length && s[left] == s[right]) {
    left--;
    right++;
  }
  return right - left - 1;
}

function longestPalindrome(s: string): string {
  if (!s) return s;

  let left = 0;
  let right = 0;

  for (let i = 0; i < s.length; i++) {
    const len1 = expandAroundCenter(s, i, i);
    const len2 = expandAroundCenter(s, i, i + 1);
    const maxLength = Math.max(len1, len2);

    if (maxLength > right - left) {
      left = i - Math.floor((maxLength - 1) / 2);
      right = i + Math.floor(maxLength / 2);
    }
  }

  return s.substring(left, right + 1);
}

File: Problems/5. Longest Palindromic Substring/longest-palindromic-substring.dart
//Solution: 207ms 141.86 MB
class Solution {
  String longestPalindrome(String s) {
    if (s.isEmpty) {
      return "";
    }

    int start = 0;
    int end = 0;

    for (int i = 0; i < s.length; i++) {
      int len1 = expandAroundCenter(s, i, i);
      int len2 = expandAroundCenter(s, i, i + 1);
      int maxLength = max(len1, len2);

      if (maxLength > end - start) {
        start = i - (maxLength - 1) ~/ 2;
        end = i + maxLength ~/ 2;
      }
    }

    return s.substring(start, end + 1);
  }

  int expandAroundCenter(String s, int left, int right) {
    while (left >= 0 && right < s.length && s[left] == s[right]) {
      left--;
      right++;
    }
    return right - left - 1;
  }
}